velox/exec/AggregateInfo.cpp

```diff
@@ -103,6 +103,8 @@ std::vector<AggregateInfo> toAggregateInfo(
         aggResultType,
         operatorCtx.driverCtx()->queryConfig());
 
+    info.function->initialize(
+        aggregate.rawInputTypes, aggResultType, info.constantInputs);
```

Comment:
Hi @liujiayi771, I'm starting reviewing this PR. I was busy with the VeloxCon and sorry for the delay. Could you remind me of the use case where you need the rawInputTypes in the function-level state? Also, I think aggResultType is already passed to info.function when it's created through Aggregate::create(). Do we still need aggResultType in the function-level state?

Reply:
Hi @kagamiori, thank you for taking the time to review this PR.

> Could you remind me of the use case where you need the rawInputTypes in the function-level state?

In the Spark decimal avg, I need to calculate the decimal type of the sum in the `intermediateType` based on the `rawInputType`, and obtain the correct precision and scale. We have discussed in https://github.com/facebookincubator/velox/discussions/9048#discussioncomment-8780398 that it is incorrect to calculate based on the `resultType`. When the precision of the `resultType` > 34, we cannot infer the precision of the sum.

> Do we still need aggResultType in the function-level state?

Can we access the `resultType` info from `aggInfo` in the `writeIntermediateResult` and `writeFinalResult` methods of `SimpleAggregateAdapter`'s struct?

Reply:
Hi @liujiayi771, thank you for reminding me of the discussion in #9048. Could you also share a concrete example about how you use the change in this PR to solve the problem of the merge_extract function of decimal avg()? I'd like to understand it better.

For example, suppose you're calling companion functions of `avg(decimal(30, 20)) -> row(decimal(38, 20), bigint) -> decimal(34, 24)`. Do you use the AggregateCompanionAdapter to generate the companion functions, and what rawInputTypes and outputType_ do you set in the AggregationNode of them?

Reply:
Hi @kagamiori. In fact, the problem of the companion function not being registered in this case has not been resolved. We still cannot use the aggregate companion function, and can only register a normal aggregate function.

However, we intend to use this normal aggregate function in Gluten. If we can obtain the `rawInputType` in the function state, we can implement the logic for Spark decimal avg based on the simple function interface. I already have a basic implementation, and the code is [here](https://github.com/liujiayi771/velox/blob/decimal-avg-test/velox/functions/sparksql/aggregates/DecimalAverageAggregate.h).

Reply:
Hi @liujiayi771, thank you for the clarification. Are you still going to use the avg_merge_extract function for decimal type and do you plan to implement it manually as an aggregation function separate from avg? I'm asking because you mentioned in https://github.com/facebookincubator/velox/discussions/9048#discussioncomment-8796536 that you plan to pass the rawInputType of avg as the input type to the factory of avg_merge_extract. I'm concerned that this does not work with the current auto-generated factory code of merge_extract companion functions because the auto-generated code uses the input type to resolve the original result type of avg.
https://github.com/facebookincubator/velox/blob/7a3607098f2202134f9c76318ade1c0dfef94595/velox/exec/AggregateCompanionAdapter.cpp#L412-L413

Reply:
@kagamiori Are you saying that, for the `merge_extract` companion function, if we follow the implementation of this PR, the `rawInputType` received in its `initialize` method (used to initialize the function state of this aggregate function) is actually the `intermediateType` of the original aggregate function?

If I understand your concern correctly, my answer is yes, for the `merge_extract` companion function, its `rawInputType` is indeed the `intermediateType` of the original aggregate function. However, this does not affect Spark decimal avg's ability to obtain its `sumType`; the `sumType` doesn't always need to be obtained from the original `rawInputType`. For the `merge_extract` companion function, its `rawInputType->childAt(0)` is the `sumType`. I can use the `initialize` method to determine whether the current initialization is for the `merge_extract` companion function based on whether `rawInputType` is a `ROW`.

Reply:
Hi @liujiayi771, thank you for the explanation. I remember the intermediate type is `ROW(DECIMAL(min(38, a_precision + 10), a_scale), bigint)` while the result type is `DECIMAL(min(38, a_precision + 4), min(38, a_scale + 4))`. As we discussed in https://github.com/facebookincubator/velox/discussions/9048, min(38, a_precision + 4) may not be inferable from min(38, a_precision + 10) if a_precision >= 28. The AggregateCompanionAdapter would require the result type be inferable from intermediate type. How would you plan to address this problem?

Reply:
Hi @kagamiori. As we discussed before, it is not possible to resolve the `resultType` through `intermediateType` in Velox. But in the case of Spark, this can be resolved through a tricky workaround, and this tricky code will only be kept in our forked repo. In Velox, the avg function signature of decimal still does not support companion aggregate functions.

The signature for decimal is
```
exec::AggregateFunctionSignatureBuilder()
        .integerVariable("a_precision")
        .integerVariable("a_scale")
        .integerVariable("i_precision", "min(38, a_precision + 10)")
        .integerVariable("r_precision", "min(38, a_precision + 4)")
        .integerVariable("r_scale", "min(38, a_scale + 4)")
        .argumentType("DECIMAL(a_precision, a_scale)")
        .intermediateType("ROW(DECIMAL(i_precision, a_scale), bigint)")
        .returnType("DECIMAL(r_precision, r_scale)")
        .build()
```
This signature will not register companion aggregate functions, because `isResultTypeResolvableGivenIntermediateType` is false. So, it does not involve resolve `resultType` from `intermediateType`, and there is no issue.

But if we still want to use the auto-generated `avg_merge_extract`, this is where the tricky part comes in. We will only add this tricky approach in our forked repo and introduce a new signature.
```
exec::AggregateFunctionSignatureBuilder()
        .integerVariable("a_precision")
        .integerVariable("a_scale")
        .argumentType("DECIMAL(a_precision, a_scale)")
        .intermediateType("ROW(DECIMAL(a_precision, a_scale), BIGINT)")
        .returnType("DECIMAL(a_precision, a_scale)")
        .build()
```

This signature does not correspond to the actual situation. Its purpose is to register companion aggregate functions for decimal type. The type of this signature is the same as the one above, but the precision and scale of decimal are incorrect. However, this situation still works fine for Spark, because Spark does not use function type resolution logic in Velox, and we have removed the resolve check for `merge_extract` function in our repo. All types are defined by the Spark plan, not based on the signature resolution by Velox. We entirely rely on the types in the Spark plan to build the vector types used in operators and functions. You can understand that Spark itself already includes the type resolution functionality, and we simply build the Velox plan and execute it based on the types in the Spark plan.

We can think of a better way to resolve the types for the `merge_extract` function of the decimal avg later, but even if we don't register the companion functions, the decimal avg still needs the information in the function state to complete the final result calculation.